Add new post on React Suspense: What it is, how to use it, gotchas, patterns and more (#738)
artsy/artsy.github.io: _posts/2022-09-08-patterns-with-suspense.markdown

one can use to get around the gotchas. Suspense is new; it's like the transition
period when Functional Components (FCs) came out and people were moving from
class based components to hooks. And remember when render props were a thing?
They were a way to use dependency injection to "inject" props into a component,
allowing us to share data dependencies across a react tree. That pattern was
cool at the time but it wasn't without its headaches, and the community hadn't
yet settled on contexts as they're now commonly understood. It was a pattern
that we had to come up with. It's the same thing with Suspense. React has
provided the tools for us to work with concurrency patterns in our UI, but since
Suspense is so new we still need to figure out just how to do it right.

The following bits of code have been made generic for demonstration purposes,
but if you use [Relay][relay] (like we do at Artsy) you're officially in a
suspense-based world and all of the following patterns will apply.

### An example

Say you've got an `<Autocomplete />` component that allows you to enter some
text which fetches data to render a dropdown list. The data is being requested

> via a hook that uses suspense, and the request is being triggered on each

Reviewer: > The data is being requested via a hook that uses suspense

**thought:** I am wondering here if the hook is using Suspense or if it's the opposite here. I was under the impression that Suspense is aware of the hook that fetches the data

Author: Technically speaking react is unaware of the hook until it throws a promise, at which point the promise is registered by react suspense. And the hook isn't doing anything _other_ than throwing a promise, which is just a js pattern. Its neatly decoupled

Reviewer: Thanks for explaining Chris 🙏 I wasn't getting it right

Author: Ah, good suggestion, will do 👍

demoed at React Conf in [2018][reactconf], has been discussed and debated [for
years][umbrella], and the first parts started to ship in [v16][v16]. And now,
with React 18 shipping to a client app near you, it's time to discuss some of
the patterns that are emerging in the community and how to avoid some common
pitfalls that you might encounter when implementing Suspense (or refactoring)
your UI in support of it.

<!-- more -->

Before we begin, it's worth noting that Suspense is a dense subject. It's more
than just a pattern for efficient data-fetching; it's a new way to think about
your UI by elevating concurrency to the status of first-class citizen. When
first trying to understand how this fits together it can feel overwhelming, and
for that reason it's important to take things slow and ideally build a simple
app that uses the pattern. At that point you'll likely experience an "ah hah!"
moment that renders the pattern clear and from which more complex UI
interactions can be designed. May this post be your guide!

## Definitions

First, what is Suspense? Quoting the React Docs:

> Suspense lets your components “wait” for something before they can render

> "Waiting" could mean showing a spinner while an API request takes place, a
> bundle split component is lazy-loaded via [`React.lazy`][splitting], or even a
> long-running, expensive computation is performed (without locking up the UI).

Reviewer 2: ```suggestion
"Waiting" could mean showing a spinner while an API request takes place, lazy-loading a
bundle split component via [`React.lazy`][splitting], or even performing a
long-running, expensive computation (without locking up the UI).
```

It's a way to declaratively state that in some fashion or another _we need to
explicitly manage time_.

To demonstrate, imagine an app that renders logged-in user info along with a
list of artists and artworks that they follow:

```tsx
const App = () => {
  return (
    <Suspense fallback="Loading...">
      <UserInfo />

      <Suspense fallback="Fetching Artists...">
        <FollowedArtists />
      </Suspense>
      <Suspense fallback="Fetching Artworks...">
        <FollowedArtworks />
      </Suspense>
    </Suspense>
  )
}
```

A couple things are happening here that are important to pay attention to, most
notably the structure of the component tree and the nesting of various Suspense
boundaries. Breaking it down a bit:

> 1. `<App />` is wrapped in a top-level `<Suspense />` boundary. This tells React

Reviewer 3: I might be just missing it, but were showing the returned result of <App /> in the example, but not that is wrapped by a <Suspense /> boundary, right? Should we say the <App /> component returns out a tree with a top-level <Suspense /> boundary?

>    that it is waiting for something to happen

Reviewer 4: ```suggestion
   that it is waiting for something to happen.
```

1. The first component nested within the top-level suspense boundary is

>    `<UserInfo />`. React attempts to render it

Reviewer 4: ```suggestion
   `<UserInfo />`. React attempts to render it.
```

> 1. Inside of `<UserInfo />` there's a mechanism for fetching user data. When
>    triggered, this tells React that we need to "suspend"

Reviewer 2: I found this is a bit confusing since we haven't seen `<UserInfo />` at this point - maybe we could just add something like this to clarify: 

```suggestion
1. Inside of `<UserInfo />` there's a mechanism for fetching user data (you'll see how
   that works below. When triggered, this tells React that we need to "suspend"
```

1. At this moment the top-level "Loading..." fallback indicator appears.
1. `<FollowedArtists />` and `<FollowedArtworks />` execute and also trigger
   fallback loading states
1. `<UserInfo />` finishes loading and reveals the page
1. `<FollowedArtists />` and `<FollowedArtworks />` render and show their
   content as soon as they finish loading

The nested suspense boundaries give us fine-grained control over time management
in our UI. We're able to say: before the user data is fetched, visually pause
everything (as user info is critical); once that's done render artists and
artworks as they are secondary, and the order in which they resolve is

> unimportant. The structure of the boundaries are like adding breakpoints to ones

Reviewer 2: ```suggestion
unimportant. The structure of the boundaries are like adding breakpoints to one's
```

code, and some breakpoints are clearly more important than others judging by the
depth in which the Suspense boundaries appear in the component tree.

The question is: how does a component tell React that it needs to suspend?

## Building your first Suspense component

This is where things get interesting and kind of weird and we discover that what
seems complex is actually deceptively simple. To instruct React to suspend, you

> simply `throw` a promise in the same way that you would `throw` an error:

Reviewer 3: Fascinating. I'm really curious about how this is working!

```tsx
const UserInfo = () => {
  throw new Promise(...)

  return (
    <div>...</div>
  )
}
```

That's Suspense's mechanism in a nutshell and all you really need to know to
understand some of the more complex ideas that follow. In a React component
render cycle, `throw promise` means: suspend.

## But why?

When I first started exploring Suspense I asked myself this question a lot,
because I was only thinking about suspense as a way to fetch data, and I had
already been fetching data with sequential loading states just fine for years.
Why is it necessary to introduce a strange new JavaScript pattern ("throwing a
promise?") for something that I can easily do in a `useEffect`?

What I didn't understand is that Suspense is designed for something other than
_fetching on render_; it's designed around the concept of _rendering as you

> fetch_. This is _the_ crucial idea to understand about Suspense: that **one

Reviewer 3: Super helpful conceptualization here.

should be able to execute an entire component tree in one go with nothing
preventing any other component from executing, regardless of whether there are

> required data dependencies needed for rendering the display**. The performance

Reviewer 3: I acknowledge I may be a bit jaded, but I have an almost knee-jerk reaction to performance claims these days and just wondering if we have anything to prove this out...?

Author: The best way to understand it is to imagine component `<App />`, which contains components <A />, which contains <B />, which contains <C />, and each have their own data dependencies. The "suspense" way is execute `<App />` which in turn fires off A, B and C concurrently. A's request might take a long time, and so we "suspend" it (and show a loading state), wheres B and C are faster. Rather than having to depend on A in order to trigger B and C -- thus creating a waterfall (imagine a crome network panel flamegraph) -- each of those requests are fired in parallel, thus speeding up the entire app. It's a remarkably faster pattern. And remember that "request" can mean anything that takes some measure of time -- a long running computation, for example.

Author: (I should probably add this as a paragraph)

implications of this are very cool and a bit mind-bending once fully understood!

> Lets demonstrate with a few examples.

Reviewer 2: ```suggestion
Let's demonstrate with a few examples.
```

because, again, data and code have been effectively separated. On `render` the
whole component tree executes and areas that need loading states simply
_display_ the loading state via `Suspense` boundary, not block execution. And if
you need formal error handling, wrap `<Suspense>` in an additional [Error
Boundary][error] for display.

Here's another pseudocode example, using a router which preloads all route data
ahead of time:

```tsx
const routes = makeRoutes([
  {
    path: '/':
    query: loadQuery(homeQuery)
    Component: () => {
      const data = usePreloadedQuery(homeQuery)
      ...
    }
  },
  {
    path: '/artists':
    query: loadQuery(artistsQuery)
    Component: () => {
      const data = usePreloadedQuery(artistsQuery)
      ...
    }
  }
])

bootApp(() => (
  <Suspense fallback="Loading route...">
    {routes}
  </Suspense>
))
```

This pattern is _flexible_. One can imagine any number of ways we could
`prepare()` this data to be `read()`. And underneath it all the mechanism is
_simple_; its just us throwing a promise to suspend, throwing an error, or
returning data to be used in the component. Very cool.

## Gotchas and Emerging Patterns Around Them

> Now that suspense is properly understood lets discuss some common patterns that

Reviewer 2: ```suggestion
Now that suspense is properly understood, let's discuss some common patterns that
```

  useEffect(() => {
    onChange(response)
  }, [response])

  return null
}
```

The `Fetcher` component doesn't actually render anything; it just fetches data
via the `useArtistQuery` hook -- and hooks, as you know, can **only** be used in
React components. Once the data has been returned it sends the response back to
the parent via the `onChange` callback prop. This then pipes the data into our
`<Autocomplete />` component.

With our flickering UI problem now solved, we're left with figuring out how to
update the `isLoading` prop on our `<Autocomplete />` component. How? We need to
create a loading toggle component that is updated every time a component
suspends and pass that in as the suspense fallback:

```tsx
<Suspense fallback={<LoadingToggle />}>
  <Fetcher />
</Suspense>

<Autocomplete ... />
```

Full example:

```tsx
const ArtistAutocomplete = () => {
  const [query, setQuery] = useState(null)
  const [options, setOptions] = useState([])
  const [isLoading, setIsLoading] = useState(false)

  const handleOnChange = (response) => {
    setIsLoading(false)
    setOptions(response)
  }

  return (
    <>
      <Suspense fallback={<LoadingToggle onChange={setIsLoading} />}>

>         <Fetcher query={query} onChange={handleResponse} />

Reviewer 2: ```suggestion
        <Fetcher query={query} onChange={handleOnChange} />
```

      </Suspense>

      <Autocomplete
        onChange={setQuery}
        options={options}
        isLoading={isLoading}
      />
    </>
  )
}

const Fetcher = ({ query, onChange }) => {
  const response = useArtistQuery(query)

  useEffect(() => {
    onChange(response)
  }, [response])

  return null
}

const LoadingToggle = ({ onChange }) => {
  useEffect(() => {
    onChange(true)
  }, [])

  return null
}
```

When the suspense boundary fallback has been triggered the `<LoadingToggle />`
component renders, and instantly we call the `onChange` callback with `true` to

> alert the parent that our component has suspended. And again, similar to
> `<Fetcher />`, we return `null` from the `LoadingToggle` because all we're
> concerned about is learning when the suspense boundary has been triggered to set
> the state in the parent, and we don't need to render visuals for the fallback;
> our aim is solely to update the `isLoading` prop on `<Autocomplete />`.

Reviewer 2: Got a bit lost reading this bit, maybe we could break it down:

```suggestion
alert the parent that our component has suspended. And again, similar to
`<Fetcher />`, we return `null` from the `LoadingToggle`. That's because all we're
concerned about is learning when the suspense boundary has been triggered: once
we know it's been triggered, we set the state in the parent. And we don't need to 
render visuals for the fallback; our aim is solely to update the `isLoading` prop on
`<Autocomplete />`.
```
